internal/component/discovery/discovery.go

```diff
@@ -110,13 +110,36 @@ type Component struct {
 	newDiscoverer chan struct{}
 
 	creator Creator
+
+	sdMetrics      discovery.DiscovererMetrics
+	refreshMetrics discovery.DiscovererMetrics
 }
 
 // New creates a discovery component given arguments and a concrete Discovery implementation function.
 func New(o component.Options, args component.Arguments, creator Creator) (*Component, error) {
+	// The metrics stay the same, even if the config of the SD mechanism changes.
+	// This means it's ok to do the registration now, with the initial config.
+	// This is similar ot how Prometheus does it.
+	discConfig := creator(args)
+	//TODO(ptodev): Change upstream Prometheus to have NewDiscovererMetrics outside of discovery.Config,
+	// to show that it's independent of the config.
+
+	//TODO(ptodev): It looks messy to have a refresh metrics separate from the other SD metrics.
+	// Can we make this cleaner? The reason it was made like this currently,
+	// is so that metrics exposed by Prometheus executables don't change.
+	refreshMetrics := discovery.NewRefreshMetrics(o.Registerer)
```

Comment:
This feels ugly, so we dont have to Register the the refreshMetrics but we do need to unregister them? And then since we can get a newdiscover on the channel we have to unregister there. The fact they aren't balanced from looking at the code makes it rough.